### ISSUE TITLE
Unable to connect to local registry on first attempt 

### ISSUE THREAD

**Issue by soniamitchell** [ScottishCovidResponse/data-registry#34]
Unable to connect to local registry on first attempt (in R). 

```
 Error in curl::curl_fetch_memory(url, handle = handle) : 
  Failed to connect to localhost port 8000: Connection refused 

Watching for file changes with StatReloader
```

Second attempt succeeds so long as it is attempted after `Watching for file changes with StatReloader` is printed on the console. 

Temporary fix: post in while loop until connection succeeds.

**Comment by nathan-cummings-ukaea** [comment]
Is R also calling the run script or do you assume the local registry is already running?

**Comment by alahiff** [comment]
I'm not able to reproduce the first attempt at connecting to the local registry failing (at least without R).

However, I do notice that the first time a local registry is started it takes a moment before it's able to start accepting connections, so if you try to connect to it too quickly you will get a `Connection refused` error.

I've updated the `run_scrc_server` script so that it blocks until the server is ready and it prints a message to stdout:
```
$ .scrc/scripts/run_scrc_server 
nohup: redirecting stderr to stdout
Local registry now accepting connections on http://localhost:8000
```
Previous the `run_scrc_server` script would exit before the server is ready.

**Comment by soniamitchell** [comment]
Fabulous, It works now! (R also calls the run script) 

However, I do get `Error starting local registry` printed on my console when calling the run script, which is slightly unnerving.

**Comment by alahiff** [comment]
You should only get that message if the local registry doesn't start within 30 seconds - for me it was taking under 1 second so I thought 30 would give lots of headroom :-)

Could you please run:
```
bash -c "echo > /dev/tcp/localhost/8000"
```
on the machine where you have the local registry running? It the server is running this should do nothing, i.e. there shouldn't be any error messages.

**Comment by soniamitchell** [comment]
Actually, strike that last comment. I restarted R and am continuing to get the error

```
Error in curl::curl_fetch_memory(url, handle = handle) : 
  Failed to connect to localhost port 8000: Connection refused 
```

Though, a short delay between running the script and interacting with the registry does help.

**Comment by soniamitchell** [comment]
No error messages when I run `bash -c "echo > /dev/tcp/localhost/8000"` 👍  only when I call the run script in R

**Comment by nathan-cummings-ukaea** [comment]
IT sounds like R is perhaps launching the run_server script as a separate process and carrying on regardless of whether the run_server script has finished? If that is the case, a short wait in the R script is the first thing that comes to mind.

**Comment by alahiff** [comment]
I didn't realise the run script was being called from R.

**Comment by nathan-cummings-ukaea** [comment]
When I ran the run script from the command line I immediately (there was no 30 second wait) got 'Error starting local registry' as well. Bu checking manually with my browser, the server was running.

**Comment by soniamitchell** [comment]
If the intention is to run everything via the commend line, then it would make more sense to call the run script via `fdp run`, which would make the issue I'm having irrelevant – @richardreeve ?

If I set a 1 second delay, everything works perfectly. 

The `Error starting local registry` appears almost instantly for me. Though it seems meaningless as I am able to interact with the registry.

**Comment by nathan-cummings-ukaea** [comment]
Yeah that's what I get too @soniamitchell

**Comment by soniamitchell** [comment]
I was thinking it might be useful to have some functionality in R that would interact directly with the registry, for example listing data products or downloading data for observation / quick analyses? Not sure if this is part of the plan though...

**Comment by soniamitchell** [comment]
Having spoken to @richardreeve, we might add functionality whereby R / the APIs might interact with the registry directly, without the CLI. So the issue remains. I've reverted back to the loop for the time being as it's much faster than adding 1 second delays everywhere. I'm sure I can find some middle ground that precludes the possibility of hanging indefinitely.

Strangely, I did find an argument, `wait`, that specifies whether the R interpreter should wait for the command to finish, or run it asynchronously. Setting this to TRUE didn't make a difference.
